File: build_data.js
const fs = require("node:fs");
const path = require("node:path");

const WORKSPACE = __dirname;
const SOURCE_CSV = String.raw`C:\Users\ASUS\Downloads\BLR_Restaurants_Clean.csv`;
const OUTPUT_FILE = path.join(WORKSPACE, "restaurants-data.js");

const AREA_ALIASES = {
  "btm layout": "BTM Layout",
  "bannerghatta road": "Bannerghatta Road",
  "brigade road": "Brigade Road",
  "indiranagar": "Indiranagar",
  "malleshwaram": "Malleshwaram",
  "electronic city": "Electronic City",
  "banashankari": "Banashankari",
  "kalyan nagar": "Kalyan Nagar",
  "bellandur": "Bellandur",
  "marathahalli": "Marathahalli",
  "basavanagudi": "Basavanagudi",
  "rajajinagar": "Rajajinagar",
  "lavelle road": "Lavelle Road",
  "jayanagar": "Jayanagar",
  "frazer town": "Frazer Town",
  "brookefield": "Brookefield",
  "old airport road": "Old Airport Road",
  "jp nagar": "JP Nagar",
  "kammanahalli": "Kammanahalli",
  "whitefield": "Whitefield",
  "church street": "Church Street",
  "new bel road": "New BEL Road",
  "sarjapur road": "Sarjapur Road",
  "mg road": "MG Road",
  "residency road": "Residency Road",
  hsr: "HSR Layout",
};

const ADDRESS_AREA_HINTS = [
  ["Church Street", "church street"],
  ["Bannerghatta Road", "bannerghatta"],
  ["HSR Layout", "hsr"],
  ["Indiranagar", "indiranagar"],
  ["Whitefield", "whitefield"],
  ["Koramangala", "koramangala"],
  ["Lavelle Road", "lavelle"],
  ["Brigade Road", "brigade"],
  ["Old Airport Road", "airport road"],
  ["Sarjapur Road", "sarjapur"],
  ["Electronic City", "electronic city"],
  ["MG Road", "mahatma gandhi rd"],
  ["MG Road", "mg road"],
  ["JP Nagar", "j p nagar"],
  ["JP Nagar", "jp nagar"],
];

const MOJIBAKE_MARKERS = ["Ã", "Â", "â", "\uFFFD"];

function parseCsv(text) {
  const rows = [];
  let row = [];
  let cell = "";
  let inQuotes = false;

  for (let i = 0; i < text.length; i += 1) {
    const char = text[i];
    const next = text[i + 1];

    if (char === '"') {
      if (inQuotes && next === '"') {
        cell += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
      continue;
    }

    if (char === "," && !inQuotes) {
      row.push(cell);
      cell = "";
      continue;
    }

    if ((char === "\n" || char === "\r") && !inQuotes) {
      if (char === "\r" && next === "\n") {
        i += 1;
      }
      row.push(cell);
      if (row.some((item) => item !== "")) {
        rows.push(row);
      }
      row = [];
      cell = "";
      continue;
    }

    cell += char;
  }

  if (cell.length || row.length) {
    row.push(cell);
    rows.push(row);
  }

  const [header, ...body] = rows;
  return body.map((items) => {
    const record = {};
    header.forEach((column, index) => {
      record[column] = items[index] ?? "";
    });
    return record;
  });
}

function repairText(value) {
  if (value == null) return "";
  const text = String(value).replace(/\ufeff/g, "").replace(/\u00a0/g, " ").trim();
  if (!text) return "";
  const candidates = [text];
  if (MOJIBAKE_MARKERS.some((marker) => text.includes(marker))) {
    try {
      candidates.push(Buffer.from(text, "latin1").toString("utf8"));
    } catch {}
  }
  return candidates
    .map((candidate) => candidate.replace(/\s+/g, " ").trim().replace(/^[,\s]+|[,\s]+$/g, ""))
    .sort((left, right) => weirdScore(left) - weirdScore(right) || right.length - left.length)[0];
}

function weirdScore(text) {
  return MOJIBAKE_MARKERS.reduce((score, marker) => score + (text.match(new RegExp(marker, "g")) || []).length, 0);
}

function parseFloatSafe(value) {
  const text = repairText(value).replace(/,/g, "");
  if (!text) return null;
  const number = Number.parseFloat(text);
  return Number.isFinite(number) ? number : null;
}

function parseIntSafe(value) {
  const number = parseFloatSafe(value);
  return number == null ? 0 : Math.round(number);
}

function clamp(number, lower, upper) {
  return Math.max(lower, Math.min(number, upper));
}

function slugify(value) {
  return repairText(value).toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "") || "restaurant";
}

function normalizeArea(area, address) {
  const source = repairText(area);
  const addressText = repairText(address).toLowerCase();

  if (source.toLowerCase().startsWith("koramangala")) return { area: "Koramangala", microArea: source };
  if (source.toLowerCase() === "hsr") return { area: "HSR Layout", microArea: "HSR" };

  const alias = AREA_ALIASES[source.toLowerCase()];
  if (alias) return { area: alias, microArea: source };

  if (source) {
    const titled = source
      .split(/\s+/)
      .map((word) => (["JP", "MG", "HSR", "BEL"].includes(word.toUpperCase()) ? word.toUpperCase() : word[0].toUpperCase() + word.slice(1).toLowerCase()))
      .join(" ");
    return { area: titled, microArea: source };
  }

  for (const [label, needle] of ADDRESS_AREA_HINTS) {
    if (addressText.includes(needle)) return { area: label, microArea: "" };
  }

  return { area: "Unknown Area", microArea: "" };
}

function splitCuisines(value) {
  return repairText(value)
    .split(",")
    .map((item) => item.replace(/\s+/g, " ").trim().replace(/^-+|-+$/g, ""))
    .filter(Boolean);
}

function normalizeBool(value) {
  const text = repairText(value).toLowerCase();
  if (["yes", "true", "1"].includes(text)) return true;
  if (["no", "false", "0"].includes(text)) return false;
  return null;
}

function categoryFromType(restaurantType, cuisine) {
  const text = `${restaurantType} ${cuisine}`.toLowerCase();
  if (/(dessert|ice cream|sweet|mithai|waffle)/.test(text)) return "dessert";
  if (/(cafe|coffee|bakery|beverage)/.test(text)) return "cafe";
  if (/(bar|pub|brew|lounge|microbrew)/.test(text)) return "bar";
  return "restaurant";
}

function firstAddressLine(value) {
  const text = repairText(value);
  return text ? text.split(",")[0].trim() : "";
}

function buildExplanation(restaurant) {
  const { rating, adjustedRating, reviewCount, hypeScore, underratedScore, reliabilityScore, tag } = restaurant;
  if (tag === "overhyped") {
    if (reliabilityScore >= 0.35) {
      return `Highly visible with ${reviewCount.toLocaleString("en-IN")} reviews, but the adjusted quality settles near ${adjustedRating.toFixed(1)}.`;
    }
    return `Visibility is outrunning quality here, so the popularity looks stronger than the food signal.`;
  }
  if (tag === "hidden-gem") {
    if (reliabilityScore >= 0.28) {
      return `Exceptional adjusted quality at ${adjustedRating.toFixed(1)} despite lower visibility makes this a strong hidden gem.`;
    }
    return `Strong quality with limited mainstream attention gives this real discovery potential.`;
  }
  if (reliabilityScore >= 0.5) {
    return `The adjusted score holds at ${adjustedRating.toFixed(1)} across ${reviewCount.toLocaleString("en-IN")} reviews, so this reads as dependable.`;
  }
  if (underratedScore >= 0.1) {
    return "Quality is quietly doing the work here even without oversized attention.";
  }
  if (hypeScore >= 0.1) {
    return `Mainstream enough to be visible, but the quality signal still holds in a reasonable range.`;
  }
  return "Balanced visibility, quality, and review confidence keep this in the worth-it zone.";
}

function deriveMoods(restaurant) {
  const moods = [];
  if (restaurant.hypeScore <= 0.18 && restaurant.reliabilityScore >= 0.3) moods.push("chill");
  if (restaurant.priceRange && restaurant.priceRange <= 500 && restaurant.underratedScore >= 0.12) moods.push("budget");
  if (restaurant.adjustedRating >= 4.2 && restaurant.reliabilityScore >= 0.35 && restaurant.priceRange >= 900) moods.push("fancy");
  if (restaurant.underratedScore >= 0.16 && restaurant.popularity <= 0.42) moods.push("experimental");
  if (restaurant.popularity >= 0.55 || restaurant.hypeScore >= 0.12) moods.push("indulgent");
  return moods;
}

function computeBounds(points) {
  const lats = points.map((point) => point.lat);
  const lngs = points.map((point) => point.lng);
  const south = Math.min(...lats) - 0.04;
  const north = Math.max(...lats) + 0.04;
  const west = Math.min(...lngs) - 0.04;
  const east = Math.max(...lngs) + 0.04;
  return {
    center: [Number(((west + east) / 2).toFixed(6)), Number(((south + north) / 2).toFixed(6))],
    southWest: [Number(west.toFixed(6)), Number(south.toFixed(6))],
    northEast: [Number(east.toFixed(6)), Number(north.toFixed(6))],
  };
}

function percentile(values, ratio) {
  if (!values.length) return 1;
  const sorted = [...values].sort((left, right) => left - right);
  const index = Math.min(sorted.length - 1, Math.max(0, Math.floor(sorted.length * ratio)));
  return sorted[index] || 1;
}

function main() {
  const csvText = fs.readFileSync(SOURCE_CSV, "utf8");
  const records = parseCsv(csvText);

  const rows = records
    .map((record, index) => {
      const name = repairText(record["Restaurant Name"]);
      if (!name) return null;
      const fullAddress = repairText(record["Full Address"]);
      const { area, microArea } = normalizeArea(record["Area"], fullAddress);
      const lat = parseFloatSafe(record["Latitude"]);
      const lng = parseFloatSafe(record["Longitude"]);
      const preciseCoordinates = Number.isFinite(lat) && Number.isFinite(lng);
      const cuisines = splitCuisines(record["Cuisines"]);
      const priceRange = parseIntSafe(record["Avg Cost (Two People ₹)"]);
      const rating = clamp(parseFloatSafe(record["Rating"]) || 0, 0, 5);
      const reviewCount = parseIntSafe(record["No. of Ratings"]);
      return {
        id: `blr-${index + 1}`,
        name,
        slug: `${slugify(name)}-${index + 1}`,
        area,
        microArea: repairText(microArea),
        lat,
        lng,
        preciseCoordinates,
        coordinateSource: preciseCoordinates ? "dataset" : "missing",
        cuisine: cuisines.join(", "),
        cuisines,
        priceRange,
        rating,
        reviewCount,
        restaurantType: repairText(record["Restaurant Type"]),
        ratingSource: repairText(record["Rating Source"]),
        zomatoRating: parseFloatSafe(record["Zomato Rating"]),
        zomatoReviewCount: parseIntSafe(record["Zomato No. of Ratings"]),
        tripadvisorRating: parseFloatSafe(record["TripAdvisor Rating"]),
        tripadvisorReviewCount: parseIntSafe(record["TripAdvisor No. of Reviews"]),
        swiggyRating: parseFloatSafe(record["Swiggy Rating"]),
        swiggyReviewCount: parseIntSafe(record["Swiggy No. of Ratings"]),
        phone: repairText(record["Phone"]),
        fullAddress,
        localAddress: firstAddressLine(fullAddress),
        description: repairText(record["Description"]),
        dietaryRestrictions: repairText(record["Dietary Restrictions"]),
        features: repairText(record["Features"]),
        mealType: repairText(record["Meal Type"]),
        timings: repairText(record["Meal Type"]),
        onlineOrder: normalizeBool(record["Online Order"]),
        tableBooking: normalizeBool(record["Table Booking"]),
        deliveryTimeMins: parseIntSafe(record["Delivery Time (Mins)"]),
        dataSources: repairText(record["Data Sources"]),
      };
    })
    .filter(Boolean);

  const areaPoints = new Map();
  for (const row of rows) {
    if (!row.preciseCoordinates) continue;
    if (!areaPoints.has(row.area)) areaPoints.set(row.area, []);
    areaPoints.get(row.area).push({ lat: row.lat, lng: row.lng });
  }

  const centroids = new Map();
  for (const [area, points] of areaPoints.entries()) {
    centroids.set(area, {
      lat: Number((points.reduce((sum, point) => sum + point.lat, 0) / points.length).toFixed(6)),
      lng: Number((points.reduce((sum, point) => sum + point.lng, 0) / points.length).toFixed(6)),
    });
  }

  for (const row of rows) {
    if (row.preciseCoordinates) continue;
    const centroid = centroids.get(row.area);
    if (!centroid) continue;
    row.lat = centroid.lat;
    row.lng = centroid.lng;
    row.coordinateSource = "area-centroid";
  }

  const maxReviewCount = rows.reduce((max, row) => Math.max(max, row.reviewCount), 1);
  const maxLog = maxReviewCount > 1 ? Math.log(maxReviewCount) : 1;
  const confidenceThreshold = percentile(
    rows
      .map((row) => row.reviewCount)
      .filter((count) => count > 0),
    0.6,
  );
  const platformAverageRating =
    rows.reduce((sum, row) => sum + (row.rating || 0), 0) / Math.max(rows.filter((row) => row.rating > 0).length, 1);

  rows.forEach((row, index) => {
    const popularity = row.reviewCount > 1 ? Math.log(row.reviewCount) / maxLog : 0;
    const adjustedRating =
      (row.reviewCount / (row.reviewCount + confidenceThreshold)) * row.rating +
      (confidenceThreshold / (row.reviewCount + confidenceThreshold)) * platformAverageRating;
    const weightedQualityRating = row.rating * 0.6 + adjustedRating * 0.4;
    const quality = clamp((weightedQualityRating - 3) / 2, 0, 1);
    const hypeScore = clamp(Math.max(popularity - quality, 0), 0, 1);
    const underratedScore = clamp(Math.max(quality - popularity, 0), 0, 1);
    const reliabilityScore = clamp(popularity, 0, 1);

    let tag = "worth-it";
    let tagLabel = "Worth It";
    const mediumReliability = reliabilityScore >= 0.35;
    const hiddenGemConfidence = row.reviewCount >= confidenceThreshold || reliabilityScore >= 0.28;
    if (hypeScore > 0.2 && quality < 0.75 && mediumReliability) {
      tag = "overhyped";
      tagLabel = "Overhyped";
    } else if (underratedScore > 0.2 && quality >= 0.75 && hiddenGemConfidence) {
      tag = "hidden-gem";
      tagLabel = "Hidden Gem";
    }

    Object.assign(row, {
      popularity: Number(popularity.toFixed(6)),
      quality: Number(quality.toFixed(6)),
      adjustedRating: Number(adjustedRating.toFixed(6)),
      weightedQualityRating: Number(weightedQualityRating.toFixed(6)),
      hypeScore: Number(hypeScore.toFixed(6)),
      underratedScore: Number(underratedScore.toFixed(6)),
      reliabilityScore: Number(reliabilityScore.toFixed(6)),
      tag,
      tagLabel,
      category: categoryFromType(row.restaurantType, row.cuisine),
      image: null,
    });

    row.moodTags = deriveMoods(row);
    row.explanation = buildExplanation(row);
    row.searchText = [
      row.name,
      row.area,
      row.microArea,
      row.restaurantType,
      row.cuisine,
      row.fullAddress,
      row.description,
      row.features,
      row.tagLabel,
      row.moodTags.join(" "),
    ]
      .filter(Boolean)
      .join(" ")
      .toLowerCase();
  });

  const cuisines = Array.from(new Set(rows.flatMap((row) => row.cuisines))).sort((left, right) => left.localeCompare(right));
  const areas = Array.from(new Set(rows.map((row) => row.area).filter((area) => area && area !== "Unknown Area"))).sort((left, right) =>
    left.localeCompare(right),
  );
  const mapped = rows.filter((row) => Number.isFinite(row.lat) && Number.isFinite(row.lng));
  const tagCounts = rows.reduce((acc, row) => {
    acc[row.tag] = (acc[row.tag] || 0) + 1;
    return acc;
  }, {});
  const areaCounts = rows.reduce((acc, row) => {
    acc[row.area] = (acc[row.area] || 0) + 1;
    return acc;
  }, {});

  const payload = {
    meta: {
      city: "Bangalore",
      sourceFile: SOURCE_CSV,
      totalRestaurants: rows.length,
      mappedRestaurants: mapped.length,
      maxReviewCount,
      platformAverageRating: Number(platformAverageRating.toFixed(6)),
      confidenceThreshold: Number(confidenceThreshold.toFixed(6)),
      cuisines,
      areas,
      tagCounts,
      areaCounts,
      mapBounds: computeBounds(mapped),
    },
    restaurants: rows,
  };

  fs.writeFileSync(OUTPUT_FILE, `window.BLRBITES_PAYLOAD = ${JSON.stringify(payload)};\n`, "utf8");

  console.log(`Built ${rows.length} restaurants`);
  console.log(`Mapped ${mapped.length} restaurants`);
  console.log(`Cuisines ${cuisines.length}`);
  console.log(`Areas ${areas.length}`);
}

main();

File: build_data.py
import csv
import json
import math
import re
from collections import Counter, defaultdict
from pathlib import Path


WORKSPACE = Path(__file__).resolve().parent
SOURCE_CSV = Path(r"C:\Users\ASUS\Downloads\BLR_Restaurants_Clean.csv")
OUTPUT_FILE = WORKSPACE / "restaurants-data.js"

AREA_ALIASES = {
    "btm layout": "BTM Layout",
    "bannerghatta road": "Bannerghatta Road",
    "brigade road": "Brigade Road",
    "indiranagar": "Indiranagar",
    "malleshwaram": "Malleshwaram",
    "electronic city": "Electronic City",
    "banashankari": "Banashankari",
    "kalyan nagar": "Kalyan Nagar",
    "bellandur": "Bellandur",
    "marathahalli": "Marathahalli",
    "basavanagudi": "Basavanagudi",
    "rajajinagar": "Rajajinagar",
    "lavelle road": "Lavelle Road",
    "jayanagar": "Jayanagar",
    "frazer town": "Frazer Town",
    "brookefield": "Brookefield",
    "old airport road": "Old Airport Road",
    "jp nagar": "JP Nagar",
    "kammanahalli": "Kammanahalli",
    "whitefield": "Whitefield",
    "church street": "Church Street",
    "new bel road": "New BEL Road",
    "sarjapur road": "Sarjapur Road",
    "mg road": "MG Road",
    "residency road": "Residency Road",
    "hsr": "HSR Layout",
}

ADDRESS_AREA_HINTS = [
    ("Church Street", "church street"),
    ("Bannerghatta Road", "bannerghatta"),
    ("HSR Layout", "hsr"),
    ("Indiranagar", "indiranagar"),
    ("Whitefield", "whitefield"),
    ("Koramangala", "koramangala"),
    ("Lavelle Road", "lavelle"),
    ("Brigade Road", "brigade"),
    ("Old Airport Road", "airport road"),
    ("Sarjapur Road", "sarjapur"),
    ("Electronic City", "electronic city"),
    ("MG Road", "mahatma gandhi rd"),
    ("MG Road", "mg road"),
    ("JP Nagar", "j p nagar"),
    ("JP Nagar", "jp nagar"),
]

MOJIBAKE_MARKERS = ("Ã", "Â", "â", "\ufffd")


def repair_text(value):
    if value is None:
        return ""
    text = str(value).replace("\ufeff", "").replace("\xa0", " ").strip()
    if not text:
        return ""
    candidates = [text]
    if any(marker in text for marker in MOJIBAKE_MARKERS):
        try:
            candidates.append(text.encode("latin1", errors="ignore").decode("utf-8", errors="ignore"))
        except Exception:
            pass
    scored = []
    for candidate in candidates:
        normalized = re.sub(r"\s+", " ", candidate).strip(" ,")
        weird = sum(normalized.count(marker) for marker in MOJIBAKE_MARKERS)
        scored.append((weird, -len(normalized), normalized))
    scored.sort()
    return scored[0][2]


def parse_float(value):
    text = repair_text(value)
    if not text:
        return None
    text = text.replace(",", "")
    try:
        return float(text)
    except ValueError:
        return None


def parse_int(value):
    number = parse_float(value)
    if number is None:
        return 0
    return int(round(number))


def slugify(text):
    cleaned = repair_text(text).lower()
    cleaned = re.sub(r"[^a-z0-9]+", "-", cleaned)
    cleaned = re.sub(r"-{2,}", "-", cleaned).strip("-")
    return cleaned or "restaurant"


def normalize_area(area, address):
    source = repair_text(area)
    address_text = repair_text(address).lower()

    if source.lower().startswith("koramangala"):
        return "Koramangala", source
    if source.lower() == "hsr":
        return "HSR Layout", "HSR"

    key = source.lower()
    if key in AREA_ALIASES:
        return AREA_ALIASES[key], source

    if source:
        titled = " ".join(word.upper() if word.upper() in {"JP", "MG", "HSR", "BEL"} else word.capitalize() for word in source.split())
        return titled, source

    for label, needle in ADDRESS_AREA_HINTS:
        if needle in address_text:
            return label, ""
    return "Unknown Area", ""


def split_cuisines(text):
    cuisines = []
    for part in repair_text(text).split(","):
        item = re.sub(r"\s+", " ", part).strip(" -")
        if item:
            cuisines.append(item)
    return cuisines


def normalize_bool(value):
    text = repair_text(value).lower()
    if text in {"yes", "true", "1"}:
        return True
    if text in {"no", "false", "0"}:
        return False
    return None


def category_from_type(restaurant_type, cuisine):
    text = f"{restaurant_type} {cuisine}".lower()
    if any(token in text for token in ("dessert", "ice cream", "sweet", "waffle", "mithai")):
        return "dessert"
    if any(token in text for token in ("cafe", "coffee", "bakery", "beverage")):
        return "cafe"
    if any(token in text for token in ("bar", "pub", "brew", "lounge", "microbrew")):
        return "bar"
    return "restaurant"


def first_address_line(address):
    cleaned = repair_text(address)
    if not cleaned:
        return ""
    return cleaned.split(",")[0].strip()


def clamp(number, lower, upper):
    return max(lower, min(number, upper))


def build_explanation(restaurant):
    rating = restaurant["rating"]
    adjusted = restaurant["adjustedRating"]
    reviews = restaurant["reviewCount"]
    hype = restaurant["hypeScore"]
    underrated = restaurant["underratedScore"]
    reliability = restaurant["reliabilityScore"]
    tag = restaurant["tag"]

    if tag == "overhyped":
        if reliability >= 0.35:
            return f"Highly visible with {reviews:,} reviews, but the adjusted quality settles near {adjusted:.1f}."
        return "Visibility is outrunning quality here, so the popularity looks stronger than the food signal."
    if tag == "hidden-gem":
        if reliability >= 0.28:
            return f"Exceptional adjusted quality at {adjusted:.1f} despite lower visibility makes this a strong hidden gem."
        return "Strong quality with limited mainstream attention gives this real discovery potential."
    if reliability >= 0.5:
        return f"The adjusted score holds at {adjusted:.1f} across {reviews:,} reviews, so this reads as dependable."
    if underrated >= 0.1:
        return "Quality is quietly doing the work here even without oversized attention."
    if hype >= 0.1:
        return "Mainstream enough to be visible, but the quality signal still holds in a reasonable range."
    return "Balanced visibility, quality, and review confidence keep this in the worth-it zone."


def derive_moods(restaurant):
    moods = []
    if restaurant["hypeScore"] <= 0.18 and restaurant["reliabilityScore"] >= 0.3:
        moods.append("chill")
    if restaurant["priceRange"] and restaurant["priceRange"] <= 500 and restaurant["underratedScore"] >= 0.12:
        moods.append("budget")
    if restaurant["adjustedRating"] >= 4.2 and restaurant["reliabilityScore"] >= 0.35 and restaurant["priceRange"] >= 900:
        moods.append("fancy")
    if restaurant["underratedScore"] >= 0.16 and restaurant["popularity"] <= 0.42:
        moods.append("experimental")
    if restaurant["popularity"] >= 0.55 or restaurant["hypeScore"] >= 0.12:
        moods.append("indulgent")
    return moods


def compute_bounds(points):
    lats = [point[0] for point in points]
    lngs = [point[1] for point in points]
    south = min(lats) - 0.04
    north = max(lats) + 0.04
    west = min(lngs) - 0.04
    east = max(lngs) + 0.04
    return {
        "center": [round((west + east) / 2, 6), round((south + north) / 2, 6)],
        "southWest": [round(west, 6), round(south, 6)],
        "northEast": [round(east, 6), round(north, 6)],
    }


def percentile(values, ratio):
    if not values:
        return 1
    ordered = sorted(values)
    index = min(len(ordered) - 1, max(0, int(len(ordered) * ratio)))
    return ordered[index] or 1


def main():
    if not SOURCE_CSV.exists():
        raise FileNotFoundError(f"CSV not found: {SOURCE_CSV}")

    rows = []
    with SOURCE_CSV.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        for index, row in enumerate(reader, start=1):
            name = repair_text(row.get("Restaurant Name"))
            if not name:
                continue

            address = repair_text(row.get("Full Address"))
            area, micro_area = normalize_area(row.get("Area"), address)
            cuisine_list = split_cuisines(row.get("Cuisines"))
            price_range = parse_int(row.get("Avg Cost (Two People ₹)"))
            rating = clamp(parse_float(row.get("Rating")) or 0, 0, 5)
            review_count = parse_int(row.get("No. of Ratings"))

            lat = parse_float(row.get("Latitude"))
            lng = parse_float(row.get("Longitude"))
            precise_coordinates = lat is not None and lng is not None

            rows.append(
                {
                    "id": f"blr-{index}",
                    "name": name,
                    "slug": f"{slugify(name)}-{index}",
                    "area": area,
                    "microArea": repair_text(micro_area),
                    "lat": lat,
                    "lng": lng,
                    "preciseCoordinates": precise_coordinates,
                    "coordinateSource": "dataset" if precise_coordinates else "missing",
                    "cuisine": ", ".join(cuisine_list),
                    "cuisines": cuisine_list,
                    "priceRange": price_range,
                    "rating": rating,
                    "reviewCount": review_count,
                    "restaurantType": repair_text(row.get("Restaurant Type")),
                    "ratingSource": repair_text(row.get("Rating Source")),
                    "zomatoRating": parse_float(row.get("Zomato Rating")),
                    "zomatoReviewCount": parse_int(row.get("Zomato No. of Ratings")),
                    "tripadvisorRating": parse_float(row.get("TripAdvisor Rating")),
                    "tripadvisorReviewCount": parse_int(row.get("TripAdvisor No. of Reviews")),
                    "swiggyRating": parse_float(row.get("Swiggy Rating")),
                    "swiggyReviewCount": parse_int(row.get("Swiggy No. of Ratings")),
                    "phone": repair_text(row.get("Phone")),
                    "fullAddress": address,
                    "localAddress": first_address_line(address),
                    "description": repair_text(row.get("Description")),
                    "dietaryRestrictions": repair_text(row.get("Dietary Restrictions")),
                    "features": repair_text(row.get("Features")),
                    "mealType": repair_text(row.get("Meal Type")),
                    "timings": repair_text(row.get("Meal Type")),
                    "onlineOrder": normalize_bool(row.get("Online Order")),
                    "tableBooking": normalize_bool(row.get("Table Booking")),
                    "deliveryTimeMins": parse_int(row.get("Delivery Time (Mins)")),
                    "dataSources": repair_text(row.get("Data Sources")),
                }
            )

    area_points = defaultdict(list)
    for row in rows:
        if row["preciseCoordinates"]:
            area_points[row["area"]].append((row["lat"], row["lng"]))

    area_centroids = {}
    for area, points in area_points.items():
        area_centroids[area] = {
            "lat": round(sum(point[0] for point in points) / len(points), 6),
            "lng": round(sum(point[1] for point in points) / len(points), 6),
        }

    for row in rows:
        if row["preciseCoordinates"]:
            continue
        centroid = area_centroids.get(row["area"])
        if centroid:
            row["lat"] = centroid["lat"]
            row["lng"] = centroid["lng"]
            row["coordinateSource"] = "area-centroid"

    max_review_count = max((row["reviewCount"] for row in rows), default=1)
    max_log = math.log(max_review_count) if max_review_count > 1 else 1

    confidence_threshold = percentile([row["reviewCount"] for row in rows if row["reviewCount"] > 0], 0.6)
    platform_average_rating = sum(row["rating"] for row in rows if row["rating"] > 0) / max(len([row for row in rows if row["rating"] > 0]), 1)

    for row in rows:
        popularity = math.log(row["reviewCount"]) / max_log if row["reviewCount"] > 1 else 0
        adjusted_rating = (
            (row["reviewCount"] / (row["reviewCount"] + confidence_threshold)) * row["rating"]
            + (confidence_threshold / (row["reviewCount"] + confidence_threshold)) * platform_average_rating
        )
        weighted_quality_rating = row["rating"] * 0.6 + adjusted_rating * 0.4
        quality = clamp((weighted_quality_rating - 3) / 2, 0, 1)
        hype_score = clamp(max(popularity - quality, 0), 0, 1)
        underrated_score = clamp(max(quality - popularity, 0), 0, 1)
        reliability_score = clamp(popularity, 0, 1)
        medium_reliability = reliability_score >= 0.35
        hidden_gem_confidence = row["reviewCount"] >= confidence_threshold or reliability_score >= 0.28

        if hype_score > 0.2 and quality < 0.75 and medium_reliability:
            tag = "overhyped"
            tag_label = "Overhyped"
        elif underrated_score > 0.2 and quality >= 0.75 and hidden_gem_confidence:
            tag = "hidden-gem"
            tag_label = "Hidden Gem"
        else:
            tag = "worth-it"
            tag_label = "Worth It"

        row["popularity"] = round(popularity, 6)
        row["quality"] = round(quality, 6)
        row["adjustedRating"] = round(adjusted_rating, 6)
        row["weightedQualityRating"] = round(weighted_quality_rating, 6)
        row["hypeScore"] = round(hype_score, 6)
        row["underratedScore"] = round(underrated_score, 6)
        row["reliabilityScore"] = round(reliability_score, 6)
        row["tag"] = tag
        row["tagLabel"] = tag_label
        row["category"] = category_from_type(row["restaurantType"], row["cuisine"])
        row["moodTags"] = derive_moods(row)
        row["explanation"] = build_explanation(row)
        row["image"] = None
        row["searchText"] = " ".join(
            filter(
                None,
                [
                    row["name"],
                    row["area"],
                    row["microArea"],
                    row["restaurantType"],
                    row["cuisine"],
                    row["fullAddress"],
                    row["description"],
                    row["features"],
                    row["tagLabel"],
                    " ".join(row["moodTags"]),
                ],
            )
        ).lower()

    cuisines = sorted({cuisine for row in rows for cuisine in row["cuisines"]})
    areas = sorted({row["area"] for row in rows if row["area"] and row["area"] != "Unknown Area"})
    mapped_points = [(row["lat"], row["lng"]) for row in rows if row["lat"] is not None and row["lng"] is not None]
    tag_counts = Counter(row["tag"] for row in rows)
    area_counts = Counter(row["area"] for row in rows)

    payload = {
        "meta": {
            "city": "Bangalore",
            "sourceFile": str(SOURCE_CSV),
            "totalRestaurants": len(rows),
            "mappedRestaurants": len(mapped_points),
            "maxReviewCount": max_review_count,
            "platformAverageRating": round(platform_average_rating, 6),
            "confidenceThreshold": round(confidence_threshold, 6),
            "cuisines": cuisines,
            "areas": areas,
            "tagCounts": dict(tag_counts),
            "areaCounts": dict(area_counts),
            "mapBounds": compute_bounds(mapped_points),
        },
        "restaurants": rows,
    }

    OUTPUT_FILE.write_text(
        "window.BLRBITES_PAYLOAD = " + json.dumps(payload, ensure_ascii=False, separators=(",", ":")) + ";\n",
        encoding="utf-8",
    )

    print(f"Built {len(rows)} restaurants")
    print(f"Mapped {len(mapped_points)} restaurants")
    print(f"Cuisines {len(cuisines)}")
    print(f"Areas {len(areas)}")


if __name__ == "__main__":
    main()
